includes/class-sensei-course-enrolment-manager.php

```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * File containing the class Sensei_Course_Enrolment_Manager.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Handles the management of course enrolment.
+ */
+class Sensei_Course_Enrolment_Manager {
+	const COURSE_ENROLMENT_SITE_SALT_OPTION = 'sensei_course_enrolment_site_salt';
+
+	/**
+	 * All course enrolment providers.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private static $enrolment_providers;
+
+	/**
+	 * Gets the descriptive name of the provider by ID.
+	 *
+	 * @param string $provider_id Unique identifier of the enrolment provider.
+	 *
+	 * @return string|false
+	 */
+	public static function get_enrolment_provider_name_by_id( $provider_id ) {
+		$provider = self::get_enrolment_provider_by_id( $provider_id );
+		if ( ! $provider ) {
+			return false;
+		}
+
+		$provider_class = get_class( $provider );
+
+		return $provider_class::get_name();
+	}
+
+	/**
+	 * Gets the enrolment provider object by its ID.
+	 *
+	 * @param string $provider_id Unique identifier of the enrolment provider.
+	 *
+	 * @return Sensei_Course_Enrolment_Provider_Interface|false
+	 */
+	public static function get_enrolment_provider_by_id( $provider_id ) {
+		$all_providers = self::get_all_enrolment_providers();
+		if ( ! isset( $all_providers[ $provider_id ] ) ) {
+			return false;
+		}
+
+		return $all_providers[ $provider_id ];
+	}
+
+	/**
+	 * Check if we should use the legacy enrolment check. Legacy check uses course
+	 * progress to determine enrolment.
+	 *
+	 * @return bool
+	 */
+	public static function use_legacy_enrolment_check() {
+		$use_legacy_enrolment_check = false;
+
+		// Check if WCPC is around but not offering enrolment providers (an old version).
+		if (
+			class_exists( '\Sensei_WC_Paid_Courses\Sensei_WC_Paid_Courses' ) &&
+			! class_exists( '\Sensei_WC_Paid_Courses\Course_Enrolment_Providers' )
+		) {
+			$use_legacy_enrolment_check = true;
+		}
+
+		return apply_filters( 'sensei_legacy_enrolment_check', $use_legacy_enrolment_check );
+	}
+
+	/**
+	 * Get an array of all the instantiated course enrolment providers.
+	 *
+	 * @return Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	public static function get_all_enrolment_providers() {
+		if ( ! isset( self::$enrolment_providers ) ) {
```

Comment:
Is this something that can be done during class initialization? 

If we do it during initilization the rest of the methods will not need to call get_all_enrolment_providers and they would be able to access the class' internal state directly. We also do not have the risk of new methods accessing uninitialized state, as self::$enrolments_providers requires call get_all_enrolment_providers in order to initialize.

Reply:
Right now we are not initializing the class. I am doing it lazily right now instead of in an initializer. I think we should use getters inside the class to allow for lazy methods. For singleton-like classes, I don't have too much of a preference for classes with static methods vs instantiated. When we add actions/filters to this class, I find it a bit easier to deal with static methods, but again--no strong preference. What would turning it into instantiated give us?

I think the only method I would like to keep static is `trigger_course_enrolment_check` to simplify usage outside of Sensei core.

Reply:
The main benefit is that we use language constructs for initialization. If we do all the initialization in the constructor then we know that any instance has a properly initialized internal state. 
Using lazy initialization is great when there is a performance benefit but otherwise it just requires from any new class method to follow the convention of calling `get_all_enrolment_providers` to access `enrolment_providers`.
Using the constructor for initialization has also the added benefit that you can just look at its signature and figure out the external dependencies of the class. It also tends to do the code more modular and with better encapsulation.

I understand though that we don't need more than one instances of this manager and that we are going to probably end up using it everywhere so ease of access might outweigh the above benefits. I'll leave the decision to you since you have deeper knowledge of the codebase :)

Reply:
Refactored in f0a88c766b88a7450fd6907b908ec2f015d26863.

I left `trigger_course_enrolment_check` static and the legacy check. The legacy check is removed in the chained manual course provider PR.

I took this opportunity to add another protection which didn't exist before. Instead of collecting in `__construct` I set to collect providers late in `init` to give plugins the opportunity to register their providers. If we access I also am throwing an `Exception` to bubble up so we never attempt enrolment checks until after initialization.